release_notes.md

```diff
@@ -1,3 +1,75 @@
+***********************************
+# Release Notes for OpenMDAO 3.31.0
+
+March 15, 2024
+
+This release includes a significant refactor that affects the calculation of relevance.
+Derivatives are no longer inherently coupled to drivers.  This makes it possible to compute derivatives of a system
+without having to add a "dummy" driver with the appropriate constraints and design variables. The SubmodelComp has been
+refactored to reflect these changes. Relevance and 'pre/opt/post' driver logic has also been refactored for more uniformity.
+
+Previously, the stalling of nonlinear solvers was detected using relative tolerance. If any subsolves had already been converged, solvers can see little-to-no improvement in the relative error, despite the absolute tolerance being fine. OpenMDAO now allows `stall_tol_type` to be `'abs'` or `'rel'` so that the user
+has more control over the detection of stalling solvers.
+
+Attempting to a assign a linesearch to a solver that doesn't support it will now result in an error.
+
+We no longer test against Python 3.7, as the python foundation no longer supports that version and some of our dependencies no longer work with it.
+
+Previously, there was no way of changing the top-level reports directory in the public API, and setting the corresponding environment variable had
+no effect after OpenMDAO was imported. The `set_reports_dir` has now been added to `openmdao.api`.
+
+Given some of the significant internal changes in this release, we tested it as thoroughly as we could. If you find
+any changes that break your models, please submit an issue so we can resolve them quickly.
+
+## New Deprecations
+
+- Renamed `list_problem_vars` to `list_driver_vars` [#3132](https://github.com/OpenMDAO/OpenMDAO/pull/3132)
+
+## Backwards Incompatible API Changes
+
```

Comment:
The automatic determination of derivative 'mode' is now done separately for the driver vs. a direct call to compute_totals or check_totals with a set of 'of' and 'wrt' variables that differ from the driver responses and design variables.  This can modify the keys in the dict return value of check_totals, where for example the dict used to contain a 'J_fwd' key because it ran in fwd mode, it might now contain a 'J_rev' key instead because the best derivative mode was determined to be 'rev' based on the relative sizes of the 'of' and 'wrt' variables passed to check_totals.